### PR TITLE
[SPARK-32705][SQL] Fix serialization issue for EmptyHashedRelation

### PR DESCRIPTION
**Pull request by leanken-zz**
### What changes were proposed in this pull request?
Currently, EmptyHashedRelation and HashedRelationWithAllNullKeys is an object, and it will cause JavaDeserialization Exception as following

```
20/08/26 11:13:30 WARN [task-result-getter-2] TaskSetManager: Lost task 34.0 in stage 57.0 (TID 18076, emr-worker-5.cluster-183257, executor 18): java.io.InvalidClassException: org.apache.spark.sql.execution.joins.EmptyHashedRelation$; no valid constructor
        at java.io.ObjectStreamClass$ExceptionInfo.newInvalidClassException(ObjectStreamClass.java:169)
        at java.io.ObjectStreamClass.checkDeserialize(ObjectStreamClass.java:874)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2042)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1572)
        at java.io.ObjectInputStream.readObject(ObjectInputStream.java:430)
        at org.apache.spark.serializer.JavaDeserializationStream.readObject(JavaSerializer.scala:76)
        at org.apache.spark.broadcast.TorrentBroadcast$.$anonfun$unBlockifyObject$4(TorrentBroadcast.scala:328)
```

This PR includes

* Using case object instead to fix serialization issue. 
* Also change EmptyHashedRelation not to extend NullAwareHashedRelation since it's already being used in other non-NAAJ joins.

### Why are the changes needed?
It will cause BHJ failed when buildSide is Empty and BHJ(NAAJ) failed when buildSide with null partition keys.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
* Existing UT.
* Run entire TPCDS for E2E coverage.

### CODE REVIEW

**Comment by leanken** [comment]
@cloud-fan Sorry for not cover this test scenario in [Spark-32290], since all UT are tested with in local mode,  it's without serialization. So in this time, run entire TPCDS 99 queries for E2E coverage and it is passed.

**Comment by cloud-fan** [comment]
ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 1]
```diff
@@ -1066,10 +1074,9 @@ private[joins] object LongHashedRelation {
 
 /**
  * Common trait with dummy implementation for NAAJ special HashedRelation
- * EmptyHashedRelation
  * HashedRelationWithAllNullKeys
  */
-trait NullAwareHashedRelation extends HashedRelation with Externalizable {
+trait NullAwareHashedRelation extends HashedRelation {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala:1079]: we don't need the trait if it only has one sub-class.

Comment (leanken) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala:1079]: OK, consider it has no other extension in future NAAJ scenario, it is reasonable to just delete it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 2]
```diff
@@ -1065,34 +1073,20 @@ private[joins] object LongHashedRelation {
 }
 
 /**
- * Common trait with dummy implementation for NAAJ special HashedRelation
- * EmptyHashedRelation
- * HashedRelationWithAllNullKeys
+ * A special HashedRelation indicates it built from a empty input:Iterator[InternalRow].
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala:1076]: `indicates it built` -> `indicating that it's built`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 3]
```diff
@@ -1102,36 +1096,33 @@ trait NullAwareHashedRelation extends HashedRelation with Externalizable {
...
-  override def readExternal(in: ObjectInput): Unit = {}
-
   override def estimatedSize: Long = 0
 }
 
 /**
- * A special HashedRelation indicates it built from a empty input:Iterator[InternalRow].
- * get & getValue will return null just like
- * empty LongHashedRelation or empty UnsafeHashedRelation does.
+ * A special HashedRelation indicates it built from a non-empty input:Iterator[InternalRow],
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala:1103]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 4]
```diff
@@ -1102,36 +1096,33 @@ trait NullAwareHashedRelation extends HashedRelation with Externalizable {
...
+ * which contains all null columns key.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala:1104]: with all the keys to be null.

**Review by cloud-fan — APPROVED** [review]

**Review by viirya — APPROVED** [review]
Looks okay. The title looks a bit weird. Can you come out a better PR title? Like "Fix serialization issue for EmptyHashedRelation"?

**Review by c21 — COMMENTED** [review]
Thanks @leanken for the fix. Just wondering is it possible to add a unit test for the issue?

**Comment by leanken** [comment]
> Thanks @leanken for the fix. Just wondering is it possible to add a unit test for the issue?

I considered adding a UT for this issue, too. But it's wrong usage about using scala object with Externalizable interface, since it is removed, I think scala case object itself can guarantee the serialization; besides, I've run the entire TPCDS just to make sure the E2E coverage, I think it should be enough, ^_^

**Comment by leanken** [comment]
@cloud-fan Test passed, ready to merge.

**Comment by c21** [comment]
> I considered adding a UT for this issue, too. But it's wrong usage about using scala object with Externalizable interface, since it is removed, I think scala case object itself can guarantee the serialization; besides, I've run the entire TPCDS just to make sure the E2E coverage, I think it should be enough, ^_^

@leanken - cool, thanks for testing this end to end. :)

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**